### PR TITLE
Add boxer claims support

### PR DESCRIPTION
**Pull request by adelinag08**
Part of https://github.com/SneaksAndData/snd-cli/issues/17
## Scope

Implemented:
 - Add requests to Boxer claim endpoints (get claims, insert claim, remove claim, add user, delete user)
 

## Checklist

- [x] GitHub issue exists for this change.
- [ ] Unit tests added and they pass.
- [x] Pylint 10.0/10.0 without bloating `.pylintrc` with exceptions.
- [ ] Review requested on `latest` commit.

### CODE REVIEW

File: esd_services_api_client/boxer/_connector.py [hunk 1]
```diff
@@ -32,11 +33,84 @@
...
+        """
+        response = self.http.delete(f"{self.base_url}/claim/{provider}/{user_id}")
+        response.raise_for_status()
+        return response
+
+    def add_claim(self, user_id: str, provider: str, claim: Claim) -> Optional[dict]:
+        """
+        Adds a new claim to an existing user_id, provider pair
+        """
+        claims = self.get_claims(user_id, provider)
```
Comment (george-zubrienko) [inline esd_services_api_client/boxer/_connector.py:86]: declare class as `@final` or use a private method

File: esd_services_api_client/boxer/_connector.py [hunk 2]
```diff
@@ -32,11 +33,84 @@
...
+        if claims is not None:
+            payload = {"operation": "Insert", "claims": claim.to_dict()}
+            payload_json = json.dumps(payload)
```
Comment (george-zubrienko) [inline esd_services_api_client/boxer/_connector.py:89]: this deserves a class I think, inheriting from DataclassesJsonMixin

File: esd_services_api_client/boxer/_connector.py [hunk 3]
```diff
@@ -32,11 +33,84 @@
...
+            return response.json()
+        return None
+
+    def remove_claim(self, user_id: str, provider: str, claim: Claim) -> Optional[dict]:
+        """
+        Removes the specified claim
+        """
+        claims = self.get_claims(user_id, provider)
+        if claims is not None:
+            payload = {"operation": "Delete", "claims": claim.to_dict()}
```
Comment (george-zubrienko) [inline esd_services_api_client/boxer/_connector.py:105]: same here

File: esd_services_api_client/boxer/_models.py [hunk 4]
```diff
@@ -21,101 +21,30 @@
...
-        return {
-            "claimType": self.claim_type,
-            "claimValue": self.claim_value,
-            "issuer": self.issuer,
-        }
+        return {self.claim_name: self.claim_value}
 
     @classmethod
-    def from_dict(cls, json_data: Dict):
+    def from_dict(cls, json_data: Dict) -> "Claim":
```
Comment (george-zubrienko) [inline esd_services_api_client/boxer/_models.py:39]: remove this and use DCJ instead

**Review by george-zubrienko — COMMENTED** [review]
A few comments

File: esd_services_api_client/boxer/README.md [hunk 5]
```diff
@@ -11,62 +11,52 @@ export BOXER_CONSUMER_ID="my_app_consumer"
...
-from esd_services_api_client.boxer import BoxerClaim
-claim = BoxerClaim(issuer="App1", claim_type="App1.CanManageConsumers", claim_value="true")
-conn = BoxerConnector(base_url="https://boxer.test.sneaksanddata.com")
-conn.push_user_claim(claim, "app1admin")
+from esd_services_api_client.boxer import select_authentication, BoxerClaimConnector, Claim
+auth = select_authentication("azuread", "test")
+conn = BoxerClaimConnector(base_url="https://boxer-claim.test.sneaksanddata.com", auth=auth)
+claim_to_insert = Claim("some-test.test.sneaksanddata.com", ".*")
+resp = conn.add_claim("adma@ecco.com", "azuread", claim_to_insert)
+print(resp)
```
[the remark this answers is not in the document]
Comment (adelinag08) [reply esd_services_api_client/boxer/README.md:37]: Modified README with output examples

File: esd_services_api_client/boxer/_connector.py (hunk 2 above)
Comment (adelinag08) [reply esd_services_api_client/boxer/_connector.py]: I added ClaimPayload class

File: esd_services_api_client/boxer/_connector.py [hunk 6]
```diff
@@ -32,11 +31,95 @@
...
+    def __init__(self, *, base_url: str, auth: Optional[BoxerTokenAuth] = None):
+        """Creates Boxer Claims connector, capable of managing claims
+        :param base_url: Base URL for Boxer Claims endpoint
+        :param auth: Boxer-based authentication
+        """
+        self._base_url = base_url
+        self.http = session_with_retries()
+        if auth and isinstance(auth, BoxerTokenAuth):
+            self.http.hooks["response"].append(auth.get_refresh_hook(self.http))
+        self.http.auth = auth
```
Comment (george-zubrienko) [inline esd_services_api_client/boxer/_connector.py:52]: these should be with `_` as well

File: esd_services_api_client/boxer/_connector.py [hunk 7]
```diff
@@ -32,11 +31,95 @@
...
+    @final
+    def get_claims(self, user_id: str, provider: str) -> Optional[Iterator[Claim]]:
+        """
+        Returns the claims assigned to the specified user_id and provider
+        """
+        response = self.http.get(f"{self._base_url}/claim/{provider}/{user_id}")
+        if response.status_code == 404:
+            return None
+        response.raise_for_status()
+        return _iterate_user_claims_response(response)
```
Comment (george-zubrienko) [inline esd_services_api_client/boxer/_connector.py:63]:
```suggestion
        return self._iterate_user_claims_response(response)
```

File: esd_services_api_client/boxer/_connector.py [hunk 8]
```diff
@@ -32,11 +31,95 @@
...
+        :param base_url: Base URL for Boxer Claims endpoint
+        :param auth: Boxer-based authentication
+        """
+        self._base_url = base_url
+        self.http = session_with_retries()
+        if auth and isinstance(auth, BoxerTokenAuth):
+            self.http.hooks["response"].append(auth.get_refresh_hook(self.http))
+        self.http.auth = auth
+
+    @final
```
Comment (george-zubrienko) [inline esd_services_api_client/boxer/_connector.py:54]: not method, class

File: esd_services_api_client/boxer/_connector.py [hunk 9]
```diff
@@ -32,11 +31,95 @@
...
+        """
+        Returns the claims assigned to the specified user_id and provider
+        """
+        response = self.http.get(f"{self._base_url}/claim/{provider}/{user_id}")
+        if response.status_code == 404:
+            return None
+        response.raise_for_status()
+        return _iterate_user_claims_response(response)
+
+    def add_user(self, user_id: str, provider: str) -> dict:
```
Comment (george-zubrienko) [inline esd_services_api_client/boxer/_connector.py:65]: on `remove_user` you have a typed response, yet here you have dict. Any reason for that?

File: esd_services_api_client/boxer/_models.py [hunk 10]
```diff
@@ -17,106 +17,34 @@
...
-        }
 
-    @classmethod
-    def from_dict(cls, json_data: Dict):
-        """Initialize from Dictionary
-        :param json_data: Dictionary
-        :return:
+    def to_custom_dict(self) -> dict:
+        """Convert to dict
+        :return: dict
```
Comment (george-zubrienko) [inline esd_services_api_client/boxer/_models.py:35]:
```suggestion
```

File: esd_services_api_client/boxer/_models.py [hunk 11]
```diff
@@ -17,106 +17,34 @@
...
         """
-        return BoxerClaim(
-            claim_type=json_data["claimType"],
-            claim_value=json_data["claimValue"],
-            issuer=json_data["issuer"],
-        )
+        return {self.claim_name: self.claim_value}
```
Comment (george-zubrienko) [inline esd_services_api_client/boxer/_models.py:37]:
```suggestion
```

File: esd_services_api_client/boxer/_models.py [hunk 12]
```diff
@@ -17,106 +17,34 @@
...
-        :param json_data: Dictionary
-        :return:
-        """
-        return GroupClaim(
-            group_name=json_data["groupName"],
-            group_claim_id=json_data["groupClaimId"],
-            claim=BoxerClaim.from_dict(json_data["claim"]),
-        )
+    operation: str
+    claims: dict
```
Comment (george-zubrienko) [inline esd_services_api_client/boxer/_models.py:47]:
```suggestion
    claims: dict
    
    def add_claim(self, claim: Claim) -> 'ClaimPayload':
        self.claims |= {claim.claim_name: claim.claim_value}
        return self
```

File: esd_services_api_client/boxer/_connector.py [hunk 13]
```diff
@@ -32,11 +31,95 @@
...
+def _prepare_claim_payload(
+    self, user_id: str, provider: str, claims: list[Claim], operation: str
+) -> Optional[str]:
+    """
+    Prepare payload for Inserting/Deleting claims
+    """
+    if self.get_claims(user_id, provider) is not None:
+        claims_dict = {
+            k: v for claim in claims for k, v in claim.to_custom_dict().items()
+        }
```
Comment (george-zubrienko) [inline esd_services_api_client/boxer/_connector.py:120]:
```suggestion
```

File: esd_services_api_client/boxer/_connector.py [hunk 14]
```diff
@@ -32,11 +31,95 @@
...
+        return ClaimPayload(operation, claims_dict).to_json()
```
Comment (george-zubrienko) [inline esd_services_api_client/boxer/_connector.py:121]:
```suggestion
        return reduce(lambda cp, claim: cp.add_claim(claim), claims).to_json()
```
smth like that?

**Review by george-zubrienko — COMMENTED** [review]
Should be good after this one

**Review by george-zubrienko — APPROVED** [review]

**Pull request merged**